span.go

```diff
@@ -225,18 +234,66 @@ func (s *Span) OperationName() string {
 	return s.operationName
 }
 
+// Retain increases object counter to increase the lifetime of the object
+func (s *Span) Retain(count ...int32) *Span {
+	counter := int32(1)
+	if len(count) > 0 && count[0] != 0 {
+		counter = count[0]
+	}
+	atomic.AddInt32(&s.retainCounter, counter)
+	return s
+}
+
+// Release decrements object counter and return to the
+// allocator manager  when counter will below zero
+func (s *Span) Release(count ...int32) {
+	counter := int32(-1)
+	if len(count) > 0 && count[0] != 0 {
+		counter = -count[0]
+	}
+
+	if atomic.AddInt32(&s.retainCounter, counter) < 0 {
+		if tr := s.tracer; tr != nil {
```

Comment:
This looks like a race condition to me: the counter can go up after the check.

Also, why do we need to be checking for `s.tracer`? It seems like it's trying to compensate for some other issue. I think  the returning to the pool should be based purely on the ref counting.